win32ss/drivers/miniport/xboxvmp/xboxvmp.c

```diff
@@ -389,29 +408,27 @@ XboxVmpMapVideoMemory(
 {
     PHYSICAL_ADDRESS FrameBuffer;
     ULONG inIoSpace = VIDEO_MEMORY_SPACE_MEMORY;
-    SYSTEM_BASIC_INFORMATION BasicInfo;
-    ULONG Length;
 
-    /* FIXME: this should probably be done differently, without native API */
     StatusBlock->Information = sizeof(VIDEO_MEMORY_INFORMATION);
 
-    FrameBuffer.u.HighPart = 0;
-    if (ZwQuerySystemInformation(SystemBasicInformation,
-                                 (PVOID)&BasicInfo,
-                                 sizeof(SYSTEM_BASIC_INFORMATION),
-                                 &Length) == NO_ERROR)
-    {
-        FrameBuffer.u.LowPart = BasicInfo.HighestPhysicalPageNumber * PAGE_SIZE;
-    }
-    else
+    DeviceExtension->VirtFrameBufferStart = VideoPortAllocateCommonBuffer(
+        DeviceExtension,
+        DeviceExtension->DmaAdapter,
+        DeviceExtension->FrameBufferLength,
+        &FrameBuffer,
+        FALSE,
+        NULL);
+
+    if (!DeviceExtension->VirtFrameBufferStart)
     {
-        ERR_(IHVVIDEO, "ZwQueryBasicInformation failed, assuming 64MB total memory\n");
-        FrameBuffer.u.LowPart = 60 * 1024 * 1024;
+        ERR_(IHVVIDEO, "VideoPortAllocateCommonBuffer failed\n");
+        return FALSE;
     }
+    VideoPortZeroMemory(DeviceExtension->VirtFrameBufferStart, DeviceExtension->FrameBufferLength);
 
     FrameBuffer.QuadPart += DeviceExtension->PhysFrameBufferStart.QuadPart;
     MapInformation->VideoRamBase = RequestedAddress->RequestedVirtualAddress;
-    MapInformation->VideoRamLength = 4 * 1024 * 1024;
+    MapInformation->VideoRamLength = DeviceExtension->FrameBufferLength;
```

Comment:
These assignments are overwritten by `VideoPortMapMemory` call

Reply:
This is input-output value, from MSDN: On input, specifies the number of bytes of video memory to map. On output, receives the size of the memory actually mapped, which might be rounded to a system-determined alignment boundary.